File: backend/controllers/dashboardController.js
const DonHang = require("../models/DonHang");
const TaiKhoan = require("../models/TaiKhoan");
const BienTheSanPham = require("../models/BienTheSanPham");
const ChiTietDonHang = require("../models/ChiTietDonHang");
const { Op } = require("sequelize");
const sequelize = require("../config/db");

const getDashboardSummary = async (req, res) => {
  try {
    const { fromDate, toDate } = req.query;

    let startRange, endRange;

    if (fromDate && toDate) {
      startRange = `${fromDate} 00:00:00`;
      endRange = `${toDate} 23:59:59`;
    } else {
      const now = new Date();
      const firstDay = new Date(now.getFullYear(), now.getMonth(), 1);
      startRange = firstDay.toISOString().slice(0, 10) + " 00:00:00";
      endRange = now.toISOString().slice(0, 10) + " 23:59:59";
    }
    

    const totalRevenue =
      (await DonHang.sum("tong_thanh_toan", {
        where: {
          trang_thai: "delivered",
          created_at: { 
            [Op.gte]: startRange,
            [Op.lte]: endRange 
          },
        },
      })) || 0;

    const newOrders = await DonHang.count({
      where: { 
        created_at: { 
          [Op.gte]: startRange,
          [Op.lte]: endRange 
        } 
      },
    });

    const totalCustomers = await TaiKhoan.count({
      where: { vai_tro: "customer" },
    });

    const totalInventory = await BienTheSanPham.sum("ton_kho");

    const recentOrdersRaw = await DonHang.findAll({
      order: [["created_at", "DESC"]],
      limit: 6,
      include: [
        {
          model: TaiKhoan,
          as: "nguoi_mua",
          attributes: ["ho_ten"],
        },
        {
          model: ChiTietDonHang,
          as: "chi_tiet",
          attributes: ["ten_sp_luc_mua"],
        },
      ],
    });

    const recentOrders = recentOrdersRaw.map((order) => {
      let productName = "Không rõ sản phẩm";

      if (order.chi_tiet && order.chi_tiet.length > 0) {
        if (order.chi_tiet.length === 1) {
          productName = order.chi_tiet[0].ten_sp_luc_mua;
        } else {
          productName = `${order.chi_tiet[0].ten_sp_luc_mua} ... (+${order.chi_tiet.length - 1} sp khác)`;
        }
      }

      return {
        id: order.ma_don_hang,
        customer: order.nguoi_mua ? order.nguoi_mua.ho_ten : "Khách vãng lai",
        product: productName,
        amount:
          new Intl.NumberFormat("vi-VN").format(order.tong_thanh_toan) + " ₫",
        status: order.trang_thai,
        date: new Date(order.created_at).toLocaleDateString("vi-VN"),
      };
    });

    const topProductsQty = await ChiTietDonHang.findAll({
      attributes: [
        "ten_sp_luc_mua",
        [sequelize.fn("SUM", sequelize.col("so_luong")), "sold_qty"],
        [sequelize.fn("SUM", sequelize.col("thanh_tien")), "total_revenue"],
      ],
      group: ["ten_sp_luc_mua"],
      order: [[sequelize.fn("SUM", sequelize.col("so_luong")), "DESC"]],
      limit: 5,
    });

    const topProductsRevenue = await ChiTietDonHang.findAll({
      attributes: [
        "ten_sp_luc_mua",
        [sequelize.fn("SUM", sequelize.col("so_luong")), "sold_qty"],
        [sequelize.fn("SUM", sequelize.col("thanh_tien")), "total_revenue"],
      ],
      group: ["ten_sp_luc_mua"],
      order: [[sequelize.fn("SUM", sequelize.col("thanh_tien")), "DESC"]],
      limit: 5,
    });

    const formatTop = (data) => {
      const maxVal = Math.max(
        ...data.map((item) => Number(item.dataValues.total_revenue)),
        1,
      );
      return data.map((item) => ({
        name: item.dataValues.ten_sp_luc_mua,
        sold: Number(item.dataValues.sold_qty),
        revenue: new Intl.NumberFormat("vi-VN").format(
          item.dataValues.total_revenue,
        ),
        pct: Math.round((Number(item.dataValues.total_revenue) / maxVal) * 100),
      }));
    };

    // Chart data range: if custom range is small (< 10 days), show day-by-day. 
    // For simplicity, we'll keep the 7-day logic but use the endRange as the reference.
    const chartEnd = new Date(`${toDate || new Date().toISOString().slice(0,10)} 23:59:59`);
    const chartStart = new Date(chartEnd);
    chartStart.setDate(chartStart.getDate() - 6);
    
    const chartStartStr = chartStart.toISOString().slice(0, 10) + " 00:00:00";
    const chartEndStr = chartEnd.toISOString().slice(0, 10) + " 23:59:59";

    const orders7Days = await DonHang.findAll({
      where: {
        trang_thai: "delivered",
        created_at: { 
          [Op.gte]: chartStartStr,
          [Op.lte]: chartEndStr 
        },
      },
      attributes: ["created_at", "tong_thanh_toan"],
    });

    const pad = (n) => n.toString().padStart(2, "0");

    const chartDataMap = {};
    for (let i = 6; i >= 0; i--) {
      const d = new Date(chartEnd);
      d.setDate(d.getDate() - i);
      const dateStr = `${pad(d.getDate())}/${pad(d.getMonth() + 1)}`;
      chartDataMap[dateStr] = { day: dateStr, value: 0, amount: "0" };
    }

    orders7Days.forEach((order) => {
      const orderDate = new Date(order.created_at);
      const dateStr = `${pad(orderDate.getDate())}/${pad(orderDate.getMonth() + 1)}`;
      if (chartDataMap[dateStr]) {
        chartDataMap[dateStr].value += Number(order.tong_thanh_toan);
      }
    });

    const chartData = Object.values(chartDataMap).map((item) => ({
      day: item.day,
      value: item.value,
      amount: item.value > 0 ? (item.value / 1000000).toFixed(1) + "tr" : "0",
    }));

    const orderStatusCount = await DonHang.findAll({
      attributes: [
        "trang_thai",
        [sequelize.fn("COUNT", sequelize.col("id")), "count"],
      ],
      where: { 
        created_at: { 
          [Op.gte]: startRange,
          [Op.lte]: endRange 
        } 
      },
      group: ["trang_thai"],
    });

    const statusColors = {
      delivered: { label: "Thành công", color: "#10b981" },
      shipping: { label: "Đang giao", color: "#3b82f6" },
      confirmed: { label: "Đang xử lý", color: "#8b5cf6" },
      pending: { label: "Chờ xử lý", color: "#f59e0b" },
      cancelled: { label: "Đã hủy", color: "#ef4444" },
      refunded: { label: "Hoàn tiền", color: "#f97316" },
    };

    let totalOrdersForPie = orderStatusCount.reduce(
      (sum, item) => sum + Number(item.dataValues.count),
      0,
    );

    const pieData = orderStatusCount.map((item) => {
      const count = Number(item.dataValues.count);
      const config = statusColors[item.trang_thai] || {
        label: item.trang_thai,
        color: "#9ca3af",
      };
      return {
        label: config.label,
        value:
          totalOrdersForPie > 0
            ? Math.round((count / totalOrdersForPie) * 100)
            : 0,
        color: config.color,
        count: count,
      };
    });

    res.status(200).json({
      success: true,
      stats: {
        totalRevenue: totalRevenue ?? 0,
        newOrders: newOrders ?? 0,
        totalCustomers: totalCustomers ?? 0,
        totalInventory: totalInventory ?? 0,
      },
      recentOrders: recentOrders,
      topProductsQty: formatTop(topProductsQty),
      topProductsRevenue: formatTop(topProductsRevenue),
      revenueChart: chartData,
      pieData: pieData,
    });
  } catch (error) {
    console.error("Lỗi getDashboardSummary:", error);
    res.status(500).json({ success: false, message: "Lỗi server" });
  }
};

module.exports = { getDashboardSummary };

File: backend-csharp/Controllers/DashboardController.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text.Json;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using WebFashion.Api.Models;

namespace WebFashion.Api.Controllers
{
    [Route("api/dashboard")]
    [ApiController]
    [Authorize]
    public class DashboardController : ControllerBase
    {
        private readonly AppDbContext _context;

        public DashboardController(AppDbContext context)
        {
            _context = context;
        }

        // GET: api/dashboard/summary
        [HttpGet("summary")]
        public async Task<IActionResult> GetDashboardSummary([FromQuery] string? fromDate, [FromQuery] string? toDate)
        {
            try
            {
                var userRole = User.FindFirst("vai_tro")?.Value;
                if (userRole != "admin")
                {
                    return StatusCode(403, new { success = false, message = "Quyền truy cập bị từ chối!" });
                }

                DateTime startRange;
                DateTime endRange;

                if (!string.IsNullOrEmpty(fromDate) && !string.IsNullOrEmpty(toDate))
                {
                    if (DateTime.TryParse(fromDate, out var fDate) && DateTime.TryParse(toDate, out var tDate))
                    {
                        startRange = fDate.Date;
                        endRange = tDate.Date.AddDays(1).AddTicks(-1);
                    }
                    else
                    {
                        return BadRequest(new { success = false, message = "Định dạng ngày không hợp lệ!" });
                    }
                }
                else
                {
                    var now = DateTime.Now;
                    startRange = new DateTime(now.Year, now.Month, 1);
                    endRange = now.Date.AddDays(1).AddTicks(-1);
                }

                // 1. Doanh thu
                decimal totalRevenue = await _context.DonHangs
                    .Where(o => o.TrangThai == "delivered" && o.CreatedAt >= startRange && o.CreatedAt <= endRange)
                    .SumAsync(o => (decimal?)o.TongThanhToan) ?? 0m;

                // 2. Đơn hàng mới
                int newOrders = await _context.DonHangs
                    .CountAsync(o => o.CreatedAt >= startRange && o.CreatedAt <= endRange);

                // 3. Số khách hàng
                int totalCustomers = await _context.TaiKhoans
                    .CountAsync(u => u.VaiTro == "customer");

                // 4. Tổng tồn kho
                int totalInventory = await _context.BienTheSanPhams
                    .SumAsync(b => (int?)b.TonKho) ?? 0;

                // 5. Đơn hàng gần đây (limit 6)
                var recentOrdersRaw = await _context.DonHangs
                    .Include(o => o.TaiKhoan)
                    .Include(o => o.ChiTietDonHangs)
                    .OrderByDescending(o => o.CreatedAt)
                    .Take(6)
                    .ToListAsync();

                var recentOrders = recentOrdersRaw.Select(order =>
                {
                    string productName = "Không rõ sản phẩm";
                    if (order.ChiTietDonHangs != null && order.ChiTietDonHangs.Count > 0)
                    {
                        var first = order.ChiTietDonHangs.First();
                        if (order.ChiTietDonHangs.Count == 1)
                        {
                            productName = first.TenSpLucMua;
                        }
                        else
                        {
                            productName = $"{first.TenSpLucMua} ... (+{order.ChiTietDonHangs.Count - 1} sp khác)";
                        }
                    }

                    return new
                    {
                        id = order.MaDonHang,
                        customer = order.TaiKhoan?.HoTen ?? "Khách vãng lai",
                        product = productName,
                        amount = order.TongThanhToan.ToString("C0", CultureInfo.GetCultureInfo("vi-VN")),
                        status = order.TrangThai,
                        date = order.CreatedAt.ToString("dd/MM/yyyy")
                    };
                }).ToList();

                // 6. Top sản phẩm bán chạy theo số lượng
                var topQtyRaw = await _context.ChiTietDonHangs
                    .GroupBy(ct => ct.TenSpLucMua)
                    .Select(g => new
                    {
                        Name = g.Key,
                        SoldQty = g.Sum(x => x.SoLuong),
                        TotalRevenue = g.Sum(x => x.ThanhTien)
                    })
                    .OrderByDescending(g => g.SoldQty)
                    .Take(5)
                    .ToListAsync();

                // 7. Top sản phẩm theo doanh thu
                var topRevRaw = await _context.ChiTietDonHangs
                    .GroupBy(ct => ct.TenSpLucMua)
                    .Select(g => new
                    {
                        Name = g.Key,
                        SoldQty = g.Sum(x => x.SoLuong),
                        TotalRevenue = g.Sum(x => x.ThanhTien)
                    })
                    .OrderByDescending(g => g.TotalRevenue)
                    .Take(5)
                    .ToListAsync();

                var maxValQty = topQtyRaw.Count > 0 ? (double)topQtyRaw.Max(x => x.TotalRevenue) : 1.0;
                if (maxValQty <= 0) maxValQty = 1.0;

                var formatTopQty = topQtyRaw.Select(item => new
                {
                    name = item.Name,
                    sold = item.SoldQty,
                    revenue = item.TotalRevenue.ToString("C0", CultureInfo.GetCultureInfo("vi-VN")).Replace("₫", "").Trim(),
                    pct = (int)Math.Round(((double)item.TotalRevenue / maxValQty) * 100)
                }).ToList();

                var maxValRev = topRevRaw.Count > 0 ? (double)topRevRaw.Max(x => x.TotalRevenue) : 1.0;
                if (maxValRev <= 0) maxValRev = 1.0;

                var formatTopRev = topRevRaw.Select(item => new
                {
                    name = item.Name,
                    sold = item.SoldQty,
                    revenue = item.TotalRevenue.ToString("C0", CultureInfo.GetCultureInfo("vi-VN")).Replace("₫", "").Trim(),
                    pct = (int)Math.Round(((double)item.TotalRevenue / maxValRev) * 100)
                }).ToList();

                // 8. Dữ liệu biểu đồ 7 ngày gần đây
                DateTime chartEnd = !string.IsNullOrEmpty(toDate) && DateTime.TryParse(toDate, out var tDt)
                    ? tDt.Date.AddDays(1).AddTicks(-1)
                    : DateTime.Now.Date.AddDays(1).AddTicks(-1);

                DateTime chartStart = chartEnd.Date.AddDays(-6);

                var orders7Days = await _context.DonHangs
                    .Where(o => o.TrangThai == "delivered" && o.CreatedAt >= chartStart && o.CreatedAt <= chartEnd)
                    .Select(o => new { o.CreatedAt, o.TongThanhToan })
                    .ToListAsync();

                var chartData = new List<object>();
                for (int i = 6; i >= 0; i--)
                {
                    var d = chartEnd.Date.AddDays(-i);
                    var dateStr = $"{d.Day:D2}/{d.Month:D2}";

                    decimal dailyValue = orders7Days
                        .Where(o => o.CreatedAt.Date == d.Date)
                        .Sum(o => o.TongThanhToan);

                    chartData.Add(new
                    {
                        day = dateStr,
                        value = (double)dailyValue,
                        amount = dailyValue > 0 ? (dailyValue / 1000000m).ToString("F1", CultureInfo.InvariantCulture) + "tr" : "0"
                    });
                }

                // 9. Dữ liệu biểu đồ tròn trạng thái đơn hàng
                var orderStatusCount = await _context.DonHangs
                    .Where(o => o.CreatedAt >= startRange && o.CreatedAt <= endRange)
                    .GroupBy(o => o.TrangThai)
                    .Select(g => new
                    {
                        TrangThai = g.Key,
                        Count = g.Count()
                    })
                    .ToListAsync();

                var statusColors = new Dictionary<string, (string Label, string Color)>
                {
                    { "delivered", ("Thành công", "#10b981") },
                    { "shipping", ("Đang giao", "#3b82f6") },
                    { "confirmed", ("Đang xử lý", "#8b5cf6") },
                    { "pending", ("Chờ xử lý", "#f59e0b") },
                    { "cancelled", ("Đã hủy", "#ef4444") },
                    { "refunded", ("Hoàn tiền", "#f97316") }
                };

                int totalOrdersForPie = orderStatusCount.Sum(x => x.Count);

                var pieData = orderStatusCount.Select(item =>
                {
                    var hasConfig = statusColors.TryGetValue(item.TrangThai, out var sc);
                    string label = hasConfig ? sc.Label : item.TrangThai;
                    string color = hasConfig ? sc.Color : "#9ca3af";

                    return new
                    {
                        label = label,
                        value = totalOrdersForPie > 0 ? (int)Math.Round(((double)item.Count / totalOrdersForPie) * 100) : 0,
                        color = color,
                        count = item.Count
                    };
                }).ToList();

                return Ok(new
                {
                    success = true,
                    stats = new
                    {
                        totalRevenue = (double)totalRevenue,
                        newOrders = newOrders,
                        totalCustomers = totalCustomers,
                        totalInventory = totalInventory
                    },
                    recentOrders = recentOrders,
                    topProductsQty = formatTopQty,
                    topProductsRevenue = formatTopRev,
                    revenueChart = chartData,
                    pieData = pieData
                });
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Lỗi getDashboardSummary: {ex.Message}");
                return StatusCode(500, new { success = false, message = "Lỗi server" });
            }
        }
    }
}
